forem/forem
# Revise skip link behaviour

## What type of PR is this? (check all applicable)


## Description

Please see the full rationale behind the change on the linked issue: https://github.com/forem/forem/issues/15006

This PR updates our 'route change' behaviour, and the way we focus the 'skip to content' link. The new pattern is:

- Route changes (i.e. user clicks on an internal link)
- A hidden span receives focus in JS (this span is not otherwise focusable)
- The route change is announced via an `aria-live` region
- The next `Tab` press makes the skip to content link visible

This PR also fixes a weird issue with the admin area skip to content link we introduced in https://github.com/forem/forem/pull/15340. Prior to this change,  I wasn't able to _see_ the skip to content link, although it was present and otherwise worked. I've pulled the relevant styles and `focus-visible` package into the admin area explicitly and it now seems to work consistently.

## Related Tickets & Documents

Closes https://github.com/forem/forem/issues/15006

## QA Instructions, Screenshots, Recordings

- Experiment with using the keyboard to navigate the app, making sure that when the page changes, your next Tab press shows you the skip to content link
- Particularly check the 'Search' page behaviour, as this is where we had problems in the past with the skip link popping up unexpectedly. You should be able to submit the search form by pressing `Enter` on either a keyboard or a mobile touch device without triggering the skip to content link visibility
- Check the admin area as well as the main app
- If you're up for it, test using a screen reader, checking that new routes are still announced as expected (there has been no change to this code but... just in case!) I've also added a recording below showing me testing this in Safari with VoiceOver (sound on):


https://user-images.githubusercontent.com/20773163/142188783-2ba50193-f6eb-4a0d-8698-6d3f1c5a80e3.mp4



### UI accessibility concerns?

This change is a result of carefully balancing some trade-offs in terms of accessibility - see the background on https://github.com/forem/forem/issues/15006

## Added/updated tests?

- [X] Yes - I actually removed a test as we no longer need to check for divergent behaviour in the search page. Now all pages behave consistently, which is likely better for accessibility and UX generally. NB: we can't test the skip link behaviour usefully in Cypress, as it does not currently allow us to simulate a `Tab` press properly


## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

      [Storybook](https://storybook.forem.com/) (for Crayons components)
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
      or in a [forem.dev](http://forem.dev) post
      replace this line with details on why this change doesn't need to be
      shared_
